Fix ETag handling for dynamic template responses

JosePineiro opened the pull request:
This PR updates AsyncStaticWebHandler::handleRequest to correctly manage caching headers when template processing introduces dynamic content. The previous implementation always set ETag and Last-Modified headers based on the underlying filesystem metadata, which was misleading when templates generated non-static responses (e.g., current time). Issue #237

The new implementation:
- Uses file-based ETag when .gz files or template callback is not present.
- Ensures gzip files generate consistent ETag values using _getEtag.
- For non-gzip files, generate ETag values ​​using the timestamp or file size.
- Returns 304 Not Modified only when If-None-Match matches the valid ETag.
- Removed use of Variable Length Array and String, ensuring the same implementation works consistently on all platforms.

This prevents browsers from incorrectly reusing cached content when template output is dynamic, ensuring correctness while still allowing efficient caching for static resources.

mathieucarbou commented on src/WebHandlers.cpp:
```diff
@@ -198,66 +207,61 @@ void AsyncStaticWebHandler::handleRequest(AsyncWebServerRequest *request) {
-#endif
+  // Get server ETag. If file is not GZ and we have a Template Processor, ETag=0 
+  char etag[9];
+  const char* tempFileName = request->_tempFile.name();
+  const size_t lenFilename = strlen(tempFileName);
+  const size_t gz_len = sizeof(T__gz) - 1;
+  if (lenFilename > sizeof(T__gz) && memcmp(tempFileName + lenFilename - gz_len, T__gz, gz_len) == 0) {
```
- `sizeof(T__gz) - 1` should be `strlen(T__gz)` I suppose ?
- event replace it by `3` in the fi condition...

JosePineiro replied:
The original code used these instructions (Variable Length Array). I've just restructured it.

I can make the change you suggest, but I've been asked in the past not to accumulate too many changes in a single PR.
Do you still want to switch to String + reserve() + heap in this PR?

JosePineiro replied:
static constexpr const char *T__gz = ".gz";
sizeof(T__gz) is 4. Please, count the chars  ('.', 'g', 'z', '\0')

I can replace by 3 or defined in literals.h as static constexpr size_t T__GZ_LEN = sizeof(T__gz) - 1;
I don't know if this will mean any improvement in speed.

what do you prefer?

mathieucarbou commented on src/WebHandlers.cpp:
```diff
@@ -198,66 +207,61 @@ void AsyncStaticWebHandler::handleRequest(AsyncWebServerRequest *request) {
+    // Use filesize-based ETag
+    size_t fileSize = request->_tempFile.size();
+    etagValue = static_cast<uint32_t>(fileSize);
+    
```
@JosePineiro : this copilot review has to be checked

mathieucarbou commented:
@JosePineiro @me-no-dev @willmmiles : FYI the previous behavior was using as an etag value the last modification time or file size to speed up file serving and avoid too many file reads, especially on concurrent requests. The drawback is that this is incorrect as per what an etag should be, and also incorrect in regard to the callback option (templating).

This PR fixes that for gz file, keeps same behavior for non gz files and considers the callback value.

That's a good fix + improvement, even if at the expense of more file reading for gz files.

willmmiles commented:
> @JosePineiro @me-no-dev @willmmiles : FYI the previous behavior was using as an etag value the last modification time or file size to speed up file serving and avoid too many file reads, especially on concurrent requests. The drawback is that this is incorrect as per what an etag should be, and also incorrect in regard to the callback option (templating).
> 
> This PR fixes that for gz file, keeps same behavior for non gz files and considers the callback value.
> 
> That's a good fix + improvement, even if at the expense of more file reading for gz files.

+1 for using internal checksums for GZ files.  A future PR could maybe generalize this for other kinds of content with some `getETagFor(file)` type function.

I'm glad to see the handling of templated content is getting some attention too.  The previous behaviour of treating it as static was definitely not ideal!

I do have some concerns though:

- This PR completely overrides any "Cache-Control" set by the user for non-templated static content.  While I agree that typical best practices with ETags would be to require an "If-None-Match" transaction to validate that things haven't changed, I can easily imagine that there could be a client application that preferred using caching to reduce the number of HTTP calls against the microcontroller.  For rarely updated content, some potential jankiness when it does change might be seen as the lesser evil than all the extra network operations all the time.  If a client had that previously configured, it will break with this PR.   Do we want to insist that ETags are the only supported cache management approach for non-templated content?

- This PR will break the semantics of `setLastModified()` when used with filesystems that don't support `getLastWrite()`.  Some clever user might have tried to use `setLastModified()` in conjunction with the templating engine -- ie. track when variables used in the template are updated, and call `setLastModified()` to apply that timestamp to the response.

I think what I'd ask for is:
- If the user has supplied a cache control policy, always send it.  If not set, for fully static content with an ETag, send `Cache-Control: none`.
- If the user has supplied a cache control policy, try to send Last-Modified.  Use the value from `setLastModified()` if set.  If not set, for static content only (gzip and otherwise!), use the filesystem timestamp.  (If no value is available, don't send, though.)

JosePineiro commented:
@willmmiles 
1 - I confirm that the idea in the future is to use the bool function AsyncWebServerRequest::_getEtag(File gzFile, char *etag) to provide the different types of eTAG
2 - I think "ETag" and "If-None-Match" are the most efficient and secure ways to control browser cache.
3 - If I'm not mistaken, "Last-Modified" and "If-Modified-Since" make exactly the same amount of http requests as "ETag" and "If-None-Match"
4 - If you agree with points 2 and 3, ETags should be the only supported cache management approach for non-template content.
5 - I haven't considered that any user has used LastModified() in conjunction with the template engine (i.e., tracking when variables used in the template are updated).
Is it okay to add _last_modified in this case?
6 - I'd like to include disabling the template processor if the files are binary (webm, jpg, avif, wolf2, etc.). What do you think?

willmmiles commented:
> I think I understand what you're saying. Suppose the user has defined Cache-Control: private, max-age=3600
> 
> We would send to the browser: Cache-Control: private, max-age=3600 ETag: "abc123"
> 
> This way, the browser won't request the page for the next hour. Once the hour has passed, the browser will send us; If-None-Match: "abc123"
> 
> If the etag remains, we will send: HTTP/1.1 304 Not Modified ETag: "abc123"
> 
> What to do if Cache-Control contains "no-store" or "immutable"? In both cases, an "ETag" is pointless, but sending it won't cause any problems.

You got it! :)

> 
> I still think that if we can use "ETag," we shouldn't send "Last-Modified". If we can't use "ETag," we should send "Last-Modified" if it's user-defined.
> 
> Does this seem like an acceptable solution to you?

I think we can do better.   If we send the user "last modified time" *as the ETag*, then we don't have to parse "If-Modified-Since".  Since ETags have no particular required format, I believe we can get away with this simplification.

So code like:

```cpp
char etag_buf[9];
char* etag = etag_buf;
const char* tempFileName = request->_tempFile.name();
const size_t lenFilename = strlen(tempFileName);

if (_last_modified.length()) {
   etag = _last_modified.c_str()
} else if (lenFilename > T__GZ_LEN && memcmp(tempFileName + lenFilename - T__GZ_LEN, T__gz, T__GZ_LEN) == 0) {
   ...
```

JosePineiro commented:
@willmmiles 
The user-defined "Cache-Control" is always sent. If not, "no-cache" is sent. Thank you for pointing out this improvement to me.

Implemented as specified in RFC 7232, section 2.4:
- SHOULD send an entity-tag validator unless it is not feasible to generate one. (In this function, whenever the file is GZ or we have not a Template Processor)
- SHOULD send a Last-Modified value if it is feasible to send one. (In this function,whenever the file is GZ or we have not a Template Processor, and the user has defined it)

In RFC 7232, section 3.3:
"A recipient MUST ignore If-Modified-Since if the request contains an If-None-Match header field"
I have implemented exactly this behavior, since not following this instruction could be considered a bug.

Please note that "entity-tag" is ALWAYS sent when "Last-Modified" is sent.
Therefore, in a modern browser capable of handling "entity-tag," I can't think of any circumstances under which "Last-Modified" would have any effect.
In any case, I've implemented the code to validate "Last-Modified" as you requested.

Note for @willmmiles: 
Please, answer to https://github.com/ESP32Async/ESPAsyncWebServer/discussions/220

willmmiles commented on src/WebHandlers.cpp:
```diff
@@ -187,75 +187,108 @@ bool AsyncStaticWebHandler::_searchFile(AsyncWebServerRequest *request, const St
   } else {
     response = new AsyncFileResponse(request->_tempFile, filename, emptyString, false, _callback);
-  }
+    if (!response) {
```
There's a regression here -- this check must also applied to the `AsyncBasicResponse` above.

willmmiles reviewed (changes requested):
A couple more little things.  Thanks for your perseverence on this, I'm excited to see this merged.

JosePineiro commented on src/WebHandlers.cpp:
```diff
@@ -187,75 +187,108 @@ bool AsyncStaticWebHandler::_searchFile(AsyncWebServerRequest *request, const St
+      }
+  }
+  // 2. Otherwise, if there is no ETag and no Template processor but we have Last-Modified and Last-Modified matches
+  else if (*etag == '\0' && _callback == nullptr && _last_modified.length() > 0 && ims && *ims && strcmp(ims, _last_modified.c_str()) == 0) {
```
Please see bug #237.
This PR attempts to fix that bug. The behavior you suggest is precisely what's causing problems for the user.
If you think the bug is incorrect, please let me know, and I'll discard this PR.

mathieucarbou commented on examples/Templates/Templates.ino:
```diff
@@ -33,6 +33,19 @@ static const char *htmlContent PROGMEM = R"(
+  uptimeInMinutes = t;
+  // Update caching header with a new value as well
+  if (uptimeHandler) {
+    uptimeHandler->setLastModified(static_cast<time_t>(t * 60));
```
I like this idea to easily let the user control the last time his template was updated

mathieucarbou commented on examples/Templates/Templates.ino:
```diff
@@ -33,6 +33,19 @@ static const char *htmlContent PROGMEM = R"(
+  uptimeInMinutes = t;
+  // Update caching header with a new value as well
+  if (uptimeHandler) {
+    uptimeHandler->setLastModified(static_cast<time_t>(t * 60));
```
I like this idea to easily let the user tell when any rendering of the template has to be done again.

But would something like this be easier to use and undertand ?

```suggestion
    uptimeHandler->clearCache();
```

Name to be determined, but the idea is to have the user call a function telling the handler that its template must be rendrered again since data has changed to the cache can be busted.

I feel this would be easier to understand than requiring the user to set a time, and decide a value ?

I guess we could se the time ourselves behind ?

willmmiles replied:
We already have `uptimeHandler->setLastModified()` with no arguments, which uses the current time from the platform.  I prefer keeping that name - I lean towards naming functions for what they do, as opposed to when they're to be called; I find it leads to more maintainable code in the long run as usages evolve.

I will update the example.

mathieucarbou reviewed (commented):
I like the new way and examples. 
Also easy to follow and understand!

I just wonder if the function call should even be simplified a bit more ;-)

mathieucarbou commented on src/WebHandlers.cpp:
```diff
@@ -186,54 +186,73 @@ bool AsyncStaticWebHandler::_searchFile(AsyncWebServerRequest *request, const St
+    notModified = true;
+  }
+  // 2. Otherwise, if there is no ETag but we have Last-Modified and Last-Modified matches
+  else if (*etag == '\0' && _last_modified.length() > 0 && request->header(T_IMS) == _last_modified) {
```
String overrides operator ==

mathieucarbou commented on examples/Templates/Templates.ino:
```diff
@@ -56,35 +69,68 @@ void setup() {
+      }
+      return emptyString;
+    })
+    .setLastModified("Sun, 28 Sep 2025 01:02:03 GMT");
```
irrelevant for example
